plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -1,69 +1,125 @@
-{-# LANGUAGE DataKinds      #-}
-{-# LANGUAGE GADTs          #-}
-{-# LANGUAGE KindSignatures #-}
-{-# LANGUAGE TypeOperators  #-}
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE GADTs                    #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeFamilies             #-}
+{-# LANGUAGE TypeOperators            #-}
 
-{-# LANGUAGE StrictData     #-}
+{-# LANGUAGE StrictData               #-}
 
 module PlutusCore.Builtin.Runtime where
 
 import PlutusPrelude
 
-import PlutusCore.Builtin.HasConstant
-import PlutusCore.Builtin.Meaning
-import PlutusCore.Builtin.TypeScheme
+import PlutusCore.Evaluation.Machine.ExBudget
+import PlutusCore.Evaluation.Machine.ExMemory
 import PlutusCore.Evaluation.Machine.Exception
 
 import Control.DeepSeq
 import Control.Lens (ix, (^?))
 import Control.Monad.Except
 import Data.Array
 import Data.Kind qualified as GHC (Type)
-import GHC.Exts (inline)
 import PlutusCore.Builtin.KnownType
 
+-- | Peano numbers. Normally called @Nat@, but that is already reserved by @base@.
+data Peano
+    = Z
+    | S Peano
+
 -- | Same as 'TypeScheme' except this one doesn't contain any evaluation-irrelevant types stuff.
-data RuntimeScheme val (args :: [GHC.Type]) res where
-    RuntimeSchemeResult
-        :: MakeKnown val res
-        => RuntimeScheme val '[] res
-    RuntimeSchemeArrow
-        :: ReadKnown val arg
-        => RuntimeScheme val args res
-        -> RuntimeScheme val (arg ': args) res
-    RuntimeSchemeAll
-        :: RuntimeScheme val args res
-        -> RuntimeScheme val args res
+-- @n@ represents the number of term-level arguments that a builtin takes.
+data RuntimeScheme n where
+    RuntimeSchemeResult :: RuntimeScheme 'Z
+    RuntimeSchemeArrow  :: RuntimeScheme n -> RuntimeScheme ('S n)
+    RuntimeSchemeAll    :: RuntimeScheme n -> RuntimeScheme n
```

Comment:
just curious, could this be used instead? 
https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/type_literals.html#computing-with-type-level-naturals

Reply:
It has horrible inference unless paired with a specific plugin. E.g. `'S n ~ 'S m` [implies](https://ideone.com/PKF7lV) `n ~ m`, but `(1 + n) ~ (1 + m)` [doesn't](https://ideone.com/Uq1vEa). Given that in our case `n` is an actual index, not just a phantom, it's best to use the unification-friendly `Peano`.

Reply:
We could probably make it work with GHC's `TypeNats`, but it would be just an additional annoyance.